packages/ai/src/workflows/controllers.ts

```diff
@@ -50,26 +44,34 @@ function handleFailedState<Shape extends IOShape>(
     );
   }
 
+  return undefined;
+}
+
+// assumes that the step is done
+function getOutputs<Shape extends IOShape, WorkflowShape extends IOShape>(
+  step: LLMStepInstance<Shape, WorkflowShape>
+): Extract<StepState<Shape>, { kind: "DONE" }>["outputs"] {
+  const state = step.getState();
   if (state.kind !== "DONE") {
-    return h.fatal(ERROR_MESSAGES.NOT_DONE(step.template.name, state.kind));
+    throw new Error(ERROR_MESSAGES.NOT_DONE(step.template.name, state.kind));
   }
-
-  return undefined;
+  return state.outputs;
 }
 
-export function fixAdjustRetryLoop<Shape extends IOShape>(
+export function getDefaultTransitionRule<Shape extends IOShape>(
```

Comment:
No more "addLinearRule", this function now returns a rule itself, and every specific workflow has a single rule.

I think this is as declarative as we can get, on this outer layer of configuring workflows, and this approach will hold as long as workflows are sequential.